File: read_data.jl
using CSV
using DataFrames
using Statistics
using Plots
using ProgressMeter
using StatsBase

"""
Read all FISH quantification data from a folder and store in a DataFrame.
This program:
1. Recursively finds all FISH quant files in the specified folder
2. Reads each file and extracts relevant data
3. Combines all data into a single DataFrame with metadata
4. Performs basic statistical analysis
5. Generates summary visualizations
"""

# Configuration
const FISH_QUANT_DIR = joinpath(@__DIR__, "fish_quant")  # Change this to your actual directory
const OUTPUT_DIR = joinpath(@__DIR__, "fish_quant_analysis")
const FILE_EXTENSIONS = [".csv", ".txt"]  # Only use CSV and text files

# Create output directory if it doesn't exist
mkpath(OUTPUT_DIR)

function find_fish_quant_files(dir_path)
    all_files = String[]
    
    println("Searching for FISH quant files in: $dir_path")
    
    # Walk through directory recursively
    for (root, dirs, files) in walkdir(dir_path)
        for file in files
            # Check if file has one of our target extensions
            if any(endswith(lowercase(file), ext) for ext in FILE_EXTENSIONS)
                push!(all_files, joinpath(root, file))
            end
        end
    end
    
    println("Found $(length(all_files)) potential FISH quant files.")
    return all_files
end

function extract_metadata_from_path(file_path)
    # Extract useful metadata from the file path
    # This function should be customized based on your specific file naming convention
    parts = splitpath(file_path)
    
    # Default metadata
    metadata = Dict(
        "file_name" => basename(file_path),
        "full_path" => file_path
    )
    
    # Extract folder names as potential metadata
    for (i, part) in enumerate(parts)
        if i < length(parts)  # Not the filename
            # Try to identify if this folder name represents metadata
            if occursin("sample", lowercase(part))
                metadata["sample"] = part
            elseif occursin("time", lowercase(part)) || occursin("hour", lowercase(part))
                metadata["time_point"] = part
            elseif occursin("treatment", lowercase(part)) || occursin("condition", lowercase(part))
                metadata["condition"] = part
            elseif occursin("replicate", lowercase(part)) || occursin("rep", lowercase(part))
                metadata["replicate"] = part
            end
        end
    end

    return metadata
end

function read_fish_quant_file(file_path)
    metadata = extract_metadata_from_path(file_path)
    
    try
        # Try to read as CSV with various possible delimiters
        for delimiter in [',', '\t', ';', ' ']
            try
                data = CSV.read(file_path, DataFrame, delim=delimiter)
                
                # If we get here, reading worked
                metadata["format"] = "delimited"
                metadata["delimiter"] = string(delimiter)
                
                # Check if we got actual data (more than 0 rows and at least 2 columns)
                if nrow(data) > 0 && ncol(data) >= 2
                    return data, metadata
                end
            catch
                # Try the next delimiter
                continue
            end
        end
        
        # If all delimiter attempts failed, try more flexible reading
        data = CSV.read(file_path, DataFrame, delim=nothing)
        metadata["format"] = "auto-detected"
        return data, metadata
        
    catch e
        @warn "Error reading file $file_path: $e"
        return nothing, metadata
    end
end

function process_fish_quant_data(data, metadata)
    # If data reading failed, return empty DataFrame with metadata
    if data === nothing
        processed = DataFrame()
        # Add metadata as a single row
        for (key, value) in metadata
            processed[!, "meta_$key"] = [value]
        end
        return processed
    end
    
    # Initialize processed DataFrame by copying the original data
    processed = copy(data)
    
    # Rename columns based on their content
    for col in names(processed)
        col_lower = lowercase(col)
        if contains(col_lower, r"cell|id")
            rename!(processed, col => "cell_id")
        elseif contains(col_lower, r"count|spot|intens|signal")
            rename!(processed, col => "value_$col")
        elseif contains(col_lower, r"pos|coord|x|y|z")
            rename!(processed, col => "pos_$col")
        else
            rename!(processed, col => "data_$col")
        end
    end
    
    # Add metadata columns - ensure we create arrays of the correct length
    n_rows = nrow(processed)
    for (key, value) in metadata
        processed[!, "meta_$key"] = fill(value, n_rows)
    end
    
    return processed
end


function (
    # Main function to run the entire workflow
    
    # 1. Find all FISH quant files
    quant_files = find_fish_quant_files(FISH_QUANT_DIR)
    
    if isempty(quant_files)
        println("No FISH quant files found. Check the directory path and file extensions.")
        return
    end
    
    # 2. Read and process each file
    all_data = DataFrame()
    
    println("Processing FISH quant files...")
    p = Progress(length(quant_files), desc="Reading files: ", dt=0.5)
    
    for file_path in quant_files
        # Read the file
        data, metadata = read_fish_quant_file(file_path)
        
        # Process the data
        processed_data = process_fish_quant_data(data, metadata)
        
        # Add to the combined DataFrame
        if nrow(processed_data) > 0
            if nrow(all_data) == 0
                all_data = processed_data
            else
                # Ensure all columns exist in both DataFrames
                for col in setdiff(names(processed_data), names(all_data))
                    all_data[!, col] = missing
                end
                for col in setdiff(names(all_data), names(processed_data))
                    processed_data[!, col] = missing
                end
                
                append!(all_data, processed_data)
            end
        end
        
        next!(p)
    end
    
    # 3. Save the combined data
    println("Saving combined data...")
    CSV.write(joinpath(OUTPUT_DIR, "all_fish_quant_data.csv"), all_data)
    
    println("\nSummary:")
    println("- Processed $(length(quant_files)) FISH quant files")
    println("- Combined data has $(nrow(all_data)) rows and $(ncol(all_data)) columns")
    println("- Results saved to $OUTPUT_DIR")
end

File: read_data.py
import os
import pandas as pd
import numpy as np
from tqdm import tqdm
import re
from pathlib import Path
import matplotlib.pyplot as plt
import seaborn as sns

"""
Read all FISH quantification data from a folder and store in a DataFrame.
This program:
1. Recursively finds all FISH quant files in the specified folder
2. Reads each file and extracts relevant data
3. Combines all data into a single DataFrame with metadata
4. Generates summary visualizations
"""

# Configuration
FISH_QUANT_DIR = os.path.join(os.path.dirname(__file__), "fish_quant")
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "fish_quant_analysis")
FILE_EXTENSIONS = [".csv", ".txt"]

# Create output directory if it doesn't exist
os.makedirs(OUTPUT_DIR, exist_ok=True)

def find_fish_quant_files(dir_path):
    """Find all FISH quantification files in the directory."""
    all_files = []
    
    print(f"Searching for FISH quant files in: {dir_path}")
    
    for root, _, files in os.walk(dir_path):
        for file in files:
            if any(file.lower().endswith(ext) for ext in FILE_EXTENSIONS):
                all_files.append(os.path.join(root, file))
    
    print(f"Found {len(all_files)} potential FISH quant files.")
    return all_files

def extract_metadata_from_path(file_path):
    """Extract metadata from file path and name."""
    filename = os.path.basename(file_path)
    
    # Default metadata
    metadata = {
        "file_name": filename,
        "full_path": file_path
    }
    
    # Extract time points from filename (e.g., -4h-, -8h-)
    time_match = re.search(r"-(\d+)h-", filename)
    if time_match:
        metadata["time_point"] = int(time_match.group(1))
    
    # Check for reinduction condition
    if "Reinduction" in filename:
        metadata["condition"] = "reinduction"
    else:
        metadata["condition"] = "primary"
    
    return metadata

def read_fish_quant_file(file_path):
    """Read and parse FISH quantification file."""
    metadata = extract_metadata_from_path(file_path)
    
    try:
        # Try different delimiters
        for delimiter in [',', '\t', ';', ' ']:
            try:
                data = pd.read_csv(file_path, delimiter=delimiter)
                if not data.empty and len(data.columns) >= 2:
                    metadata["format"] = "delimited"
                    metadata["delimiter"] = delimiter
                    return data, metadata
            except:
                continue
        
        # Fallback to automatic delimiter detection
        data = pd.read_csv(file_path, sep=None, engine='python')
        metadata["format"] = "auto-detected"
        return data, metadata
        
    except Exception as e:
        print(f"Error reading file {file_path}: {e}")
        return None, metadata

def process_fish_quant_data(data, metadata):
    """Process the FISH quantification data and add metadata."""
    if data is None:
        return pd.DataFrame()
    
    # Initialize processed DataFrame
    processed = data.copy()
    
    # Add time point and condition columns
    processed['time_point'] = metadata.get('time_point')
    processed['condition'] = metadata.get('condition', 'unknown')
    
    # Rename data columns based on their content
    rename_dict = {}
    for col in processed.columns:
        if col not in ['time_point', 'condition']:
            col_lower = col.lower()
            if any(term in col_lower for term in ['count', 'spot', 'intens', 'signal']):
                rename_dict[col] = f"value_{col}"
    
    processed.rename(columns=rename_dict, inplace=True)
    return processed

def main():
    """Main function to run the data processing pipeline."""
    # Find all FISH quant files
    quant_files = find_fish_quant_files(FISH_QUANT_DIR)
    
    if not quant_files:
        print(f"No FISH quant files found in {FISH_QUANT_DIR}")
        return pd.DataFrame()
    
    # Process files with progress bar
    all_data = pd.DataFrame()
    
    for file_path in tqdm(quant_files, desc="Reading files"):
        data, metadata = read_fish_quant_file(file_path)
        processed_data = process_fish_quant_data(data, metadata)
        
        if not processed_data.empty:
            all_data = pd.concat([all_data, processed_data], ignore_index=True)
    
    # Save results grouped by time point
    if not all_data.empty:
        time_points = sorted(all_data['time_point'].dropna().unique())
        conditions = all_data['condition'].unique()
        
        # Save data for each time point separately
        for tp in time_points:
            for cond in conditions:
                subset = all_data[
                    (all_data['time_point'] == tp) & 
                    (all_data['condition'] == cond)
                ]
                
                # Group by cell and region, sum values to get counts
                grouped = subset.groupby(['cell', 'frame', 'type']).size().reset_index(name='count')
                
                # Save the count data to CSV file
                output_path = os.path.join(OUTPUT_DIR, f"counts_{int(tp)}h_{cond}.csv")
                grouped.to_csv(output_path, index=False)
                
                # Plot histogram of counts for each value column (using Transcription Site data)
                value_columns = grouped
                
                if not value_columns.empty and not subset.empty:
                    # Create histogram
                    print(f"Creating histogram for {int(tp)}h ({cond})")
                    fig, ax = plt.subplots(figsize=(10, 6))
                    
                    sns.histplot(value_columns, 
                                 binwidth=1, 
                                 alpha=0.7,
                                 stat="density",
                                 label=f"Counts at {int(tp)}h ({cond})")
                    
                    ax.set_title(f"Histogram of FISH counts at {int(tp)}h ({cond})")
                    ax.set_xlabel("Count value")
                    ax.set_ylabel("Frequency")
                    ax.legend()
                    ax.set_xlim(0,200)
                    
                    # Save the histogram
                    plt.tight_layout()
                    plt.savefig(os.path.join(OUTPUT_DIR, f"histogram/{cond}_{int(tp)}h.png"))
                    plt.close()
        
        print("\nData grouped by time points:")
        for tp in time_points:
            n_primary = len(all_data[
                (all_data['time_point'] == tp) & 
                (all_data['condition'] == 'primary')
            ])
            n_reind = len(all_data[
                (all_data['time_point'] == tp) & 
                (all_data['condition'] == 'reinduction')
            ])
            print(f"- {int(tp)}h: Primary={n_primary}, Reinduction={n_reind} samples")
    
    return all_data

if __name__ == "__main__":
    all_data = main()